### DISCUSSION TITLE
Cookbooks Release v5.0.0 ArcGIS Server "WebContextURL" defined twice

### DISCUSSION THREAD

**Discussion by J-J-E** [Esri/arcgis-cookbook#389] [Q&A]
I am wondering if this is a necessity or an oversight, as the resource is defined twice at 

[arcgis][server][webcontexturl]

and 

[arcgis][server][system_properties][webcontexturl]

https://github.com/Esri/arcgis-cookbook/blob/main/templates/arcgis-enterprise-base/11.3/linux/arcgis-enterprise-primary.json

Can one safely be removed?

**Comment by cameronkroeker** [comment]
Hi @J-J-E,

Great question! They can both be safely removed from your json configuration file so as long as your are satisfied with the default values.

The `[arcgis][server][webcontexturl]` is used by the federation recipe to set the federated services url.

https://github.com/Esri/arcgis-cookbook/tree/main/templates/arcgis-enterprise-base/11.3#arcgis-enterprise-primary

```
arcgis.server.web_context_url - ArcGIS Server web context URL that will be used for the federated server services URL
```

https://github.com/Esri/arcgis-cookbook/blob/c92c7768637d5e5e2bb633a867107b8e9c206953/cookbooks/arcgis-enterprise/recipes/federation.rb#L39

https://github.com/Esri/arcgis-cookbook/blob/c92c7768637d5e5e2bb633a867107b8e9c206953/cookbooks/arcgis-enterprise/recipes/federation.rb#L53

**If removed it will default to using Web Adaptor URL:**
https://github.com/Esri/arcgis-cookbook/blob/c92c7768637d5e5e2bb633a867107b8e9c206953/cookbooks/arcgis-enterprise/attributes/server.rb#L47

The `[arcgis][server][system_properties][webcontexturl]` is used to defined the webcontexturl within the arcgis server [system properties](https://developers.arcgis.com/rest/enterprise-administration/server/serverproperties/).

https://github.com/Esri/arcgis-cookbook/tree/main/templates/arcgis-enterprise-base/11.3#arcgis-enterprise-primary

```
arcgis.server.system_properties.WebContextURL - ArcGIS Server web context URL used if there is a reverse proxy and\or load balancer
```

https://github.com/Esri/arcgis-cookbook/blob/c92c7768637d5e5e2bb633a867107b8e9c206953/cookbooks/arcgis-enterprise/recipes/server.rb#L145

If removed it will not set system_properties:
https://github.com/Esri/arcgis-cookbook/blob/c92c7768637d5e5e2bb633a867107b8e9c206953/cookbooks/arcgis-enterprise/attributes/server.rb#L66

Thanks,
Cameron K.

**Reply by J-J-E** [reply]
Got it, thanks!

**Comment by J-J-E** [comment]
@cameronkroeker based on your above response, I assumed that removing the web context URL settings from the json would set my default web context url as https://**domain/hostname**/server and not https://**ip-addres**/server. But my services published to portal carry the internal Ip address instead of the external domain name in their service url, using the below configuration.

Additionally, with the below configuration, the server private url and server url seem to have been reversed within the server admin security page

**Server Security Settings page below**
{
  "portalUrl": "https://gis-test.filmla.com/arcgis",
  "privatePortalUrl": "https://gis-test-enterprise.us-west2-c.c.ops-gcp.internal:7443/arcgis",
  "portalSecretKey": "--------",
  "portalMode": "ARCGIS_PORTAL_FEDERATION",
  "serverId": "--------",
  **"serverUrl": "https://10.168.0.23/server",**
  **"privateHostingServerUrl": "https://gis-test.filmla.com/server"**
}


**JSON config below**
"server": {
            "hostname": "gis-test-enterprise.us-west2-c.c.ops-gcp.internal",
            "configure_autostart": true,
            "install_dir": "/opt",
	    "private_url": "https://gis-test-enterprise.us-west2-c.c.ops-gcp.internal:6443/arcgis",
            "admin_username": "--------",
            "admin_password": "--------",
            "authorization_file": "--------",
            "directories_root": "/opt/arcgis/gisdata/arcgisserver",
            "log_dir": "/opt/arcgis/server/usr/logs",
            "log_level": "DEBUG",
            "config_store_type": "FILESYSTEM",
            "config_store_connection_string": "/opt/arcgis/gisdata/arcgisserver/config-store",
            "install_system_requirements": true,
            "wa_name": "server",
            "services_dir_enabled": true
        },


"portal": {
    "install_dir": "/opt",
    "hostname": "gis-test-enterprise.us-west2-c.c.ops-gcp.internal",
    "preferredidentifier": "hostname",
    "admin_username": "--------",
    "admin_password": "--------",
    "admin_email": "--------",
    "admin_full_name": "--------",
    "admin_description": "Initial account administrator",
    "security_question": "1",
    "security_question_answer": "--------",
    "log_dir": "/opt/arcgis/portal/usr/arcgisportal/logs",
    "log_level": "DEBUG",
    "enable_debug": false,
    "content_store_type": "fileStore",
    "content_store_provider": "FileSystem",
    "content_store_connection_string": "/opt/arcgis/gisdata/arcgisportal/content",
    "object_store": "",
    "authorization_file": "--------",
    "user_license_type_id": "creatorUT",
    "install_system_requirements": true,
    "wa_name": "arcgis",
    "system_properties": {
        "privatePortalURL": "https://gis-test-enterprise.us-west2-c.c.ops-gcp.internal:7443/arcgis",
        "WebContextURL": "https://gis-test.filmla.com/arcgis"
    }
},



With the above json configuration, I would have thought that:

1. Our server private url would be the hostname of the machine, not the external dns
2. Our server url would have been the external dns entry similar to portal, and not the internal Ip address. Especially since we designated using the hostname and not the ip as the identifier, so at the very least it should have been the internal hostname of the machine and not the internal ip
3. Our services registered in portal would have picked up the external dns url and web adaptor for the service url, instead of the server internal IP

@cameronkroeker can you please advise the appropriate server json configuration parameters i need so that the services and server security are properly set?

**Comment by cameronkroeker** [comment]
Hi @J-J-E,

The following 2 attributes are used for federation to determine the federated services url and federated server admin url:

- node['arcgis']['server']['web_context_url']
- node['arcgis']['server']['private_url']

https://github.com/Esri/arcgis-cookbook/blob/c92c7768637d5e5e2bb633a867107b8e9c206953/cookbooks/arcgis-enterprise/recipes/federation.rb#L39-L40

Perhaps this example is will do the trick?

```json
"server": {
    "hostname": "gis-test-enterprise.us-west2-c.c.ops-gcp.internal",
    "configure_autostart": true,
    "install_dir": "/opt",
    "web_context_url": "https://gis-test.filmla.com/server",
    "private_url": "https://gis-test-enterprise.us-west2-c.c.ops-gcp.internal:6443/arcgis",
    "admin_username": "--------",
    "admin_password": "--------",
    "authorization_file": "--------",
    "directories_root": "/opt/arcgis/gisdata/arcgisserver",
    "log_dir": "/opt/arcgis/server/usr/logs",
    "log_level": "DEBUG",
    "config_store_type": "FILESYSTEM",
    "config_store_connection_string": "/opt/arcgis/gisdata/arcgisserver/config-store",
    "install_system_requirements": true,
    "wa_name": "server",
    "services_dir_enabled": true,
    "system_properties": {
        "WebContextURL": "https://gis-test.filmla.com/server"
    }
}
```
